// styled-jsx/dist/lib/style-transform.js
'use strict';

var Stylis = require('stylis');

var stylis = new Stylis();

function disableNestingPlugin() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  var context = args[0],
      _args$ = args[3],
      parent = _args$ === undefined ? [] : _args$,
      line = args[4],
      column = args[5];

  if (context === 2) {
    parent = parent[0];
    if (typeof parent === 'string' && parent.trim().length > 0 && parent.charAt(0) !== '@') {
      throw new Error('Nesting detected at ' + line + ':' + column + '. ' + 'Unfortunately nesting is not supported by styled-jsx.');
    }
  }
}

var generator = void 0;
var filename = void 0;
var offset = void 0;

function sourceMapsPlugin() {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  var context = args[0],
      line = args[4],
      column = args[5],
      length = args[6];

  // Pre-processed, init source map

  if (context === -1 && generator !== undefined) {
    generator.addMapping({
      generated: {
        line: 1,
        column: 0
      },
      source: filename,
      original: offset
    });

    return;
  }

  // Post-processed
  if (context === -2 && generator !== undefined) {
    generator = undefined;
    offset = undefined;
    filename = undefined;

    return;
  }

  // Selector/property, update source map
  if ((context === 1 || context === 2) && generator !== undefined) {
    generator.addMapping({
      generated: {
        line: 1,
        column: length
      },
      source: filename,
      original: {
        line: line + offset.line,
        column: column + offset.column
      }
    });
  }
}

/**
 * splitRulesPlugin
 * Used to split a blob of css into an array of rules
 * that can inserted via sheet.insertRule
 *
 * courtesy of and (c) the emotion folks (with some fixes from us)
 * https://github.com/emotion-js/emotion/blob/994ea265cf5a411c5fa9b606dd140ce66776d1db/packages/emotion/src/index.js#L23-L60
 */
var isSplitRulesEnabled = false;
var splitRules = [];
var splitRulesQueue = [];
var scopeHash = void 0;
var nestedAtRules = ['m', 's', 'd', 'k', '-'];

function splitRulesPlugin(context, content, selectors, parent, line, column, length, id) {
  if (context === -2) {
    splitRules = splitRules.concat(splitRulesQueue);
    splitRulesQueue = [];
    return;
  }

  if (context === 1) {
    if (content.charAt(0) === '@') {
      splitRulesQueue.push(content);
      return '';
    }
  }

  if (context === 2) {
    if (id === 0) {
      var joinedSelectors = selectors.join(',');
      var rule = joinedSelectors + '{' + content + '}';
      if (parent.join(',') === joinedSelectors || parent[0] === '') {
        splitRulesQueue.push(rule);
      } else {
        splitRulesQueue.unshift(rule);
      }
    }
    return;
  }

  // after an at rule block
  if (context === 3) {
    var selectrs = selectors.join(',');
    if (nestedAtRules.indexOf(selectrs.charAt(1)) === -1) {
      splitRulesQueue.push('' + selectrs + content);
    } else if (selectrs.charAt(1) === 'k') {
      var animationName = scopeHash.replace(/^\./, '-');
      splitRulesQueue.push('@-webkit-' + selectrs.slice(1) + animationName + '{' + content + '}');
      splitRulesQueue.push('' + selectrs + animationName + '{' + content + '}');
    } else {
      splitRulesQueue.push(selectrs + '{' + content + '}');
    }
  }
}

stylis.use(disableNestingPlugin);
stylis.use(sourceMapsPlugin);
stylis.use(splitRulesPlugin);
stylis.set({
  cascade: false,
  compress: true
});

/**
 * Public transform function
 *
 * @param {String} hash
 * @param {String} styles
 * @param {Object} settings
 * @return {string}
 */
function transform(hash, styles) {
  var settings = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  generator = settings.generator;
  offset = settings.offset;
  filename = settings.filename;
  isSplitRulesEnabled = settings.splitRules;
  splitRules = [];
  scopeHash = hash;

  var cssString = stylis(scopeHash, styles);

  if (isSplitRulesEnabled) {
    return splitRules;
  }
  return cssString;
}

module.exports = transform;
